tests/test_distance_rows.py

```diff
@@ -0,0 +1,167 @@
+            sem=BoundedSemaphore(20)
+        )
+
+    assert res.isnull().sum().sum() == 0
+
+    # origin-major order
+    np.testing.assert_array_almost_equal(
+        res.index.values.tolist(),
+        FULL_INDEX
+    )
+
+    np.testing.assert_array_almost_equal(
+        res.meters.values,
+        FULL_RESULTS
+    )
+
+
+@pytest.mark.asyncio
+async def test_odd(mock_server, test_client_config):
+    mock_server.reset()
+    client = await AsyncDispatcher.init(test_client_config)
+
+    idx = create_dist_index(ORIGS, DESTS)
+
+    # take even rows only
+    missing = idx.iloc[range(0, len(idx), 2)]
+
+    async with aiohttp.ClientSession() as session:
+        res = await client.distance_rows(
+            missing=missing,
+            provider='google',
+            session=session,
+            # TODO: use default sem for async calls also
+            sem=BoundedSemaphore(20)
+        )
+
+    assert res.isnull().sum().sum() == 0
+
+    np.testing.assert_array_almost_equal(
+        res.index.values.tolist(),
+        [idx for i, idx in enumerate(FULL_INDEX) if i % 2 == 0]
+    )
+
+    np.testing.assert_array_almost_equal(
+        res.meters.values,
+        [idx for i, idx in enumerate(FULL_RESULTS) if i % 2 == 0]
+    )
+
+
+@pytest.mark.asyncio
+async def test_thin(mock_server, test_client_config):
+    mock_server.reset()
+    client = await AsyncDispatcher.init(test_client_config)
+
+    # take only last 2 destinations, 3 x 2 query
+    idx = create_dist_index(ORIGS, DESTS[[1, 2], :])
+    expected = [
+        1, 4, 7,    # destination-major order, dest1
+        2, 5, 8     # dest2
+    ]
+
+    async with aiohttp.ClientSession() as session:
+        res = await client.distance_rows(
+            missing=idx,
+            provider='google',
+            session=session,
+            # TODO: use default sem for async calls also
+            sem=BoundedSemaphore(20)
+        )
+
+    assert res.isnull().sum().sum() == 0
+
+    np.testing.assert_array_almost_equal(
+        res.index.values.tolist(),
+        FULL_INDEX[expected, :]
+    )
+
+    np.testing.assert_array_almost_equal(
+        res.meters.values,
+        FULL_RESULTS[expected]
+    )
+
+
+@pytest.mark.asyncio
+async def test_jumbled(mock_server, test_client_config):
+    mock_server.reset()
+    client = await AsyncDispatcher.init(test_client_config)
+
+    jumble = [8, 0, 7, 1, 6, 2, 5, 3]  # omit 4
+    idx = create_dist_index(ORIGS, DESTS).take(jumble)
+
+    # should be ordered origin-major, grouped by first incidence each origin value
+    expected = [
+        8, 7, 6,
+        0, 1, 2,
+        5, 3
+    ]
+
+    async with aiohttp.ClientSession() as session:
+        res = await client.distance_rows(
+            missing=idx,
+            provider='google',
+            session=session,
+            # TODO: use default sem for async calls also
+            sem=BoundedSemaphore(20)
+        )
+
+    assert res.isnull().sum().sum() == 0
```

Comment:
I'm actually surprised there's not a single call for this, such that you have to sum in both directions.

Reply:
There is for numpy arrays, not pandas because it's not a common use case to sum across columns probably.